cowprotocol/services
# Use Enso tx simulator for gas estimates

# Description
This PR implements an additional transaction simulator for accurate gas estimates (in order to properly score solutions) using the [Enso Transaction Simulator](https://github.com/EnsoFinance/transaction-simulator)

# Changes
- Implement a new subtype in the simulator infra module
- Since the transaction simulator cannot compute access lists it also has a web3 simulator which is used for this purpose
- Allow to configure either Tenderly or Enso simulator (if neither is configured we fall back to using Web3)

## How to test
Run a driver with a valid enso tx simulator URL configured together with autopilot, orderbook, CoW Swap UI and a baseline solver. Place an order, observe enso tx simulator being called. Also observe

```
driver::infra::observe: scoring settlement solutions={Id(0)} gas=Gas { estimate: Gas(227542), limit: Gas(273050), price: FeePerGas(Ether(119538777501)) }
```

And see that this is the same value as simulating in Tenderly
<img width="1370" alt="image" src="https://github.com/cowprotocol/services/assets/1200333/7ece4e45-9f80-4aad-aefa-6cd7c2b13787">


## Related Issues
Fixes #1897
